presto-main/src/main/java/com/facebook/presto/execution/buffer/OutputBufferMemoryManager.java

```diff
@@ -63,53 +66,79 @@ public OutputBufferMemoryManager(long maxBufferedBytes, Supplier<LocalMemoryCont
         this.maxBufferedBytes = maxBufferedBytes;
         this.systemMemoryContextSupplier = Suppliers.memoize(systemMemoryContextSupplier::get);
         this.notificationExecutor = requireNonNull(notificationExecutor, "notificationExecutor is null");
-
-        bufferBlockedFuture = SettableFuture.create();
-        bufferBlockedFuture.set(null);
     }
 
-    public synchronized void updateMemoryUsage(long bytesAdded)
+    public void updateMemoryUsage(long bytesAdded)
     {
-        Optional<LocalMemoryContext> systemMemoryContext = getSystemMemoryContext();
-
-        // If closed is true, that means the task is completed. In that state,
-        // the output buffers already ignore the newly added pages, and therefore
-        // we can also safely ignore any calls after OutputBufferMemoryManager is closed.
-        // If the systemMemoryContext doesn't exist, the task is probably already
-        // aborted, so we can just return (see the comment in getSystemMemoryContext()).
-        if (closed || !systemMemoryContext.isPresent()) {
+        LocalMemoryContext systemMemoryContext = getSystemMemoryContextOrNull();
+        if (systemMemoryContext == null) {
+            // If the systemMemoryContext doesn't exist, the task is probably already
+            // aborted, so we can just return (see the comment in getSystemMemoryContextOrNull()).
             return;
         }
 
-        long currentBufferedBytes = bufferedBytes.addAndGet(bytesAdded);
+        ListenableFuture<?> waitForMemory = null;
+        SettableFuture<?> notifyUnblocked = null;
+        long currentBufferedBytes;
+        synchronized (this) {
+            // If closed is true, that means the task is completed. In that state,
+            // the output buffers already ignore the newly added pages, and therefore
+            // we can also safely ignore any calls after OutputBufferMemoryManager is closed.
+            if (closed) {
+                return;
+            }
+
+            currentBufferedBytes = bufferedBytes.addAndGet(bytesAdded);
+            ListenableFuture<?> blockedOnMemory = systemMemoryContext.setBytes(currentBufferedBytes);
+            if (!blockedOnMemory.isDone()) {
+                if (this.blockedOnMemory != blockedOnMemory) {
+                    this.blockedOnMemory = blockedOnMemory;
+                    waitForMemory = blockedOnMemory; // only register a callback when blocked and the future is different
```

Comment:
can you explain this logic?  What does it mean when the future is different?

Reply:
If the future is different it means that the memory reservation returned a new, not yet complete listener and we need to register a callback for when memory is available. If memory is available or it's not available but we've already registered a callback on the current instance- there's no need to register one. The previous implementation could attach arbitrarily many callbacks to the same future as well as callbacks to "not blocked" futures which dramatically increased contention on the output buffer itself.

Summary: only register a callback when we're waiting on memory and only do that one time per `ListenableFuture<?>` instance.